File: library/src/main/java/wannabe/grid/VisibilityGrid.java
package wannabe.grid;

import java.util.Iterator;

import wannabe.Voxel;
import wannabe.grid.iterators.EmptyIterator;

/** A Grid that can hide or show its voxels. */
public class VisibilityGrid implements Grid {
  private static final AllNeighbors NO_NEIGHBORS = new AllNeighbors();
  
  private final String name;
  private final Grid source;

  private boolean visible = true;
  private boolean dirty = true;

  public VisibilityGrid(String name, Grid source) {
    this.name = name;
    this.source = source;
  }

  @Override public Iterator<Voxel> iterator() {
    dirty = false;

    return visible ? source.iterator() : new EmptyIterator();
  }

  @Override public boolean isDirty() {
    return realIsDirty();
  }

  @Override public int size() {
    return visible ? source.size() : 0;
  }

  public void show() {
    dirty = !visible;
    visible = true;
  }

  public void hide() {
    dirty = visible;
    visible = false;
  }

  /** Reverses visibility state. */
  public void toggle() {
    dirty = true;
    visible = !visible;
  }

  private boolean realIsDirty() {
    return dirty || source.isDirty();
  }

  @Override public AllNeighbors neighbors(Voxel voxel) {
    return visible ? source.neighbors(voxel) : NO_NEIGHBORS;
  } 

  @Override public String toString() {
    return name + "; visible? " + visible + " (size: " + size() + ")";
  }
}

File: library/src/main/java/wannabe/grid/VisibilityGrid.kt
package wannabe.grid

import wannabe.Voxel
import wannabe.grid.iterators.EmptyIterator

/** A Grid that can hide or show its voxels.  */
class VisibilityGrid(
  private val name: String,
  private val source: Grid
) : Grid {
  private var visible = true
  private var dirty = true

  override fun iterator(): Iterator<Voxel> {
    dirty = false
    return if (visible) source.iterator() else EmptyIterator()
  }

  override val isDirty: Boolean
    get() = realIsDirty()

  override val size: Int
    get() = if (visible) source.size else 0

  fun show() {
    dirty = !visible
    visible = true
  }

  fun hide() {
    dirty = visible
    visible = false
  }

  /** Reverses visibility state.  */
  fun toggle() {
    dirty = true
    visible = !visible
  }

  private fun realIsDirty(): Boolean {
    return dirty || source.isDirty
  }

  override fun neighbors(voxel: Voxel): AllNeighbors {
    return if (visible) source.neighbors(voxel) else NO_NEIGHBORS
  }

  override fun toString(): String {
    return "$name; visible? $visible (size: $size)"
  }

  companion object {
    private val NO_NEIGHBORS = AllNeighbors()
  }
}
